accelerator/iri_proxy_apis.c

```diff
@@ -0,0 +1,420 @@
+/*
+ * Copyright (C) 2018-2019 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#include "iri_proxy_apis.h"
+#include "utils/handles/lock.h"
+
+#define IRI_PROXY_APIS_LOGGER "iri_proxy_apis"
+
+static logger_id_t logger_id;
+lock_handle_t cjson_lock;
+
+void iri_proxy_apis_logger_init() { logger_id = logger_helper_enable(IRI_PROXY_APIS_LOGGER, LOGGER_DEBUG, true); }
+
+int iri_proxy_apis_logger_release() {
+  logger_helper_release(logger_id);
+  if (logger_helper_destroy() != RC_OK) {
+    ta_log_critical("Destroying logger failed %s.\n", IRI_PROXY_APIS_LOGGER);
+    return EXIT_FAILURE;
+  }
+
+  return 0;
+}
+
+status_t iri_proxy_apis_lock_init() {
+  if (lock_handle_init(&cjson_lock)) {
+    return SC_CONF_LOCK_INIT;
+  }
+  return SC_OK;
+}
+
+status_t iri_proxy_apis_lock_destroy() {
+  if (lock_handle_destroy(&cjson_lock)) {
+    return SC_CONF_LOCK_DESTROY;
+  }
+  return SC_OK;
+}
+
+status_t api_attach_to_tangle(const iota_client_service_t* const service, const char* const obj, char** json_result) {
+  status_t ret = SC_OK;
+  attach_to_tangle_req_t* req = attach_to_tangle_req_new();
+  attach_to_tangle_res_t* res = attach_to_tangle_res_new();
+  char_buffer_t* res_buff = char_buffer_new();
+  if (req == NULL || res == NULL || res_buff == NULL) {
+    ret = SC_TA_OOM;
+    ta_log_error("%s\n", "SC_TA_OOM");
+    goto done;
+  }
+
+  lock_handle_lock(&cjson_lock);
+  if (service->serializer.vtable.attach_to_tangle_deserialize_request(obj, req) != RC_OK) {
+    lock_handle_unlock(&cjson_lock);
+    ret = SC_CCLIENT_JSON_PARSE;
+    ta_log_error("%s\n", "SC_CCLIENT_JSON_PARSE");
+    goto done;
+  }
+  lock_handle_unlock(&cjson_lock);
+
+  lock_handle_lock(&cjson_lock);
+  if (iota_client_attach_to_tangle(service, req, res) != RC_OK) {
+    lock_handle_unlock(&cjson_lock);
+    ret = SC_CCLIENT_FAILED_RESPONSE;
+    ta_log_error("%s\n", "SC_CCLIENT_FAILED_RESPONSE");
+    goto done;
+  }
+  lock_handle_unlock(&cjson_lock);
+
+  if (service->serializer.vtable.attach_to_tangle_serialize_response(res, res_buff) != RC_OK) {
+    ret = SC_CCLIENT_JSON_PARSE;
+    ta_log_error("%s\n", "SC_CCLIENT_JSON_PARSE");
+    goto done;
+  }
+
+  *json_result = (char*)malloc((res_buff->length + 1) * sizeof(char));
```

Comment:
How about letting us declare a new function which takes the responsibility to cast a `char_buff_t` into `string`?
This function can be used in every proxy API

Reply:
Copy that!